packages/zoe/src/contractFacet/seat.js

```diff
@@ -0,0 +1,107 @@
+// @ts-check
+
+import { E } from '@agoric/eventual-send';
+import { assert, details } from '@agoric/assert';
+
+import { isOfferSafe } from './offerSafety';
+
+import '../../exported';
+import '../internal-types';
+
+/** @type {MakeZcfSeatAdminKit} */
+export const makeZcfSeatAdminKit = (
+  allSeatStagings,
+  zoeSeatAdmin,
+  seatData,
+  getAmountMath,
+) => {
+  // The proposal and notifier are not reassigned.
+  const { proposal, notifier } = seatData;
+
+  // The currentAllocation, exited, and stagedAllocation may be reassigned.
+  let currentAllocation = harden(seatData.initialAllocation);
+  let exited = false; // seat is "active"
+  const assertExitedFalse = () => assert(!exited, `seat has been exited`);
+
+  /** @type {ZCFSeatAdmin} */
+  const zcfSeatAdmin = harden({
+    commit: seatStaging => {
+      assertExitedFalse();
+      assert(
+        allSeatStagings.has(seatStaging),
+        details`The seatStaging ${seatStaging} was not recognized`,
+      );
+      currentAllocation = seatStaging.getStagedAllocation();
+      E(zoeSeatAdmin).replaceAllocation(currentAllocation);
+    },
+    updateHasExited: () => {
+      assertExitedFalse();
+      exited = true;
+    },
+  });
+
+  /** @type {ZCFSeat} */
+  const zcfSeat = harden({
+    exit: () => {
+      assertExitedFalse();
+      zcfSeatAdmin.updateHasExited();
+      E(zoeSeatAdmin).exit();
+    },
+    kickOut: (msg = 'Kicked out of seat') => {
+      assertExitedFalse();
+      zcfSeat.exit();
+      assert.fail(msg);
+    },
```

Comment:
IIUC correctly the purpose of `exit` vs `kickOut`, then `exit` should terminate the notifier with `updater.finish` whereas `kickOut` should should terminate it with `updater.fail`.